README.md
# Cyli-

Optimizing storage of data cylinders through the use of reinforcement learning.



history.txt

0,0,0
1,5,0
2,0,0
3,0,0
4,0,0
5,6,0
6,7,0
7,0,0
8,10,0
9,0,0
0,0,1
1,3,1
2,3,1
3,0,1
4,0,1
5,8,1
6,1,1
7,0,1
8,4,1
9,0,1
0,0,2
1,9,2
2,1,2
3,0,2
4,0,2
5,1,2
6,5,2
7,0,2
8,4,2
9,0,2
0,0,3
1,6,3
2,2,3
3,0,3
4,0,3
5,6,3
6,4,3
7,0,3
8,8,3
9,0,3
0,0,4
1,10,4
2,10,4
3,0,4
4,0,4
5,9,4
6,9,4
7,0,4
8,2,4
9,0,4
0,0,5
1,1,5
2,6,5
3,0,5
4,0,5
5,4,5
6,4,5
7,0,5
8,5,5
9,0,5
0,0,6
1,0,6
2,0,6
3,0,6
4,0,6
5,9,6
6,2,6
7,0,6
8,2,6
9,0,6
0,0,7
1,5,7
2,1,7
3,0,7
4,0,7
5,5,7
6,4,7
7,0,7
8,2,7
9,0,7
0,0,8
1,4,8
2,0,8
3,0,8
4,0,8
5,0,8
6,0,8
7,0,8
8,7,8
9,0,8
0,0,9
1,9,9
2,10,9
3,0,9
4,0,9
5,10,9
6,8,9
7,0,9
8,10,9
9,0,9
0,0,10
1,2,10
2,2,10
3,0,10
4,0,10
5,1,10
6,9,10
7,0,10
8,3,10
9,0,10
0,0,11
1,7,11
2,8,11
3,0,11
4,0,11
5,3,11
6,2,11
7,0,11
8,3,11
9,0,11
0,0,12
1,3,12
2,3,12
3,0,12
4,0,12
5,8,12
6,4,12
7,0,12
8,3,12
9,0,12
0,0,13
1,9,13
2,9,13
3,0,13
4,0,13
5,1,13
6,4,13
7,0,13
8,2,13
9,0,13
0,0,14
1,10,14
2,3,14
3,0,14
4,0,14
5,4,14
6,6,14
7,0,14
8,2,14
9,0,14
0,0,15
1,9,15
2,10,15
3,0,15
4,0,15
5,10,15
6,9,15
7,0,15
8,4,15
9,0,15
0,0,16
1,3,16
2,6,16
3,0,16
4,0,16
5,2,16
6,6,16
7,0,16
8,5,16
9,0,16
0,0,17
1,9,17
2,0,17
3,0,17
4,0,17
5,2,17
6,2,17
7,0,17
8,3,17
9,0,17
0,0,18
1,2,18
2,8,18
3,0,18
4,0,18
5,0,18
6,9,18
7,0,18
8,7,18
9,0,18
0,0,19
1,4,19
2,0,19
3,0,19
4,0,19
5,7,19
6,10,19
7,0,19
8,9,19
9,0,19
0,0,20
1,8,20
2,2,20
3,0,20
4,0,20
5,4,20
6,2,20
7,0,20
8,10,20
9,0,20
0,0,21
1,7,21
2,0,21
3,0,21
4,0,21
5,5,21
6,0,21
7,0,21
8,0,21
9,0,21
0,0,22
1,2,22
2,9,22
3,0,22
4,0,22
5,2,22
6,8,22
7,0,22
8,5,22
9,0,22
0,0,23
1,8,23
2,3,23
3,0,23
4,0,23
5,10,23
6,3,23
7,0,23
8,8,23
9,0,23
0,0,24
1,5,24
2,7,24
3,0,24
4,0,24
5,1,24
6,6,24
7,0,24
8,2,24
9,0,24
0,0,25
1,9,25
2,10,25
3,0,25
4,0,25
5,4,25
6,3,25
7,0,25
8,8,25
9,0,25
0,0,26
1,7,26
2,4,26
3,0,26
4,0,26
5,1,26
6,3,26
7,0,26
8,0,26
9,0,26
0,0,27
1,6,27
2,3,27
3,0,27
4,0,27
5,9,27
6,9,27
7,0,27
8,8,27
9,0,27
0,0,28
1,10,28
2,8,28
3,0,28
4,0,28
5,0,28
6,4,28
7,0,28
8,7,28
9,0,28
0,0,29
1,4,29
2,7,29
3,0,29
4,0,29
5,8,29
6,8,29
7,0,29
8,3,29
9,0,29
0,0,30
1,6,30
2,10,30
3,0,30
4,0,30
5,9,30
6,1,30
7,0,30
8,1,30
9,0,30
0,0,31
1,7,31
2,4,31
3,0,31
4,0,31
5,6,31
6,1,31
7,0,31
8,4,31
9,0,31
0,0,32
1,4,32
2,2,32
3,0,32
4,0,32
5,4,32
6,0,32
7,0,32
8,7,32
9,0,32
0,0,33
1,0,33
2,1,33
3,0,33
4,0,33
5,0,33
6,10,33
7,0,33
8,8,33
9,0,33
0,0,34
1,7,34
2,3,34
3,0,34
4,0,34
5,3,34
6,1,34
7,0,34
8,0,34
9,0,34
0,0,35
1,9,35
2,5,35
3,0,35
4,0,35
5,8,35
6,0,35
7,0,35
8,9,35
9,0,35
0,0,36
1,7,36
2,10,36
3,0,36
4,0,36
5,3,36
6,7,36
7,0,36
8,0,36
9,0,36
0,0,37
1,0,37
2,3,37
3,0,37
4,0,37
5,10,37
6,3,37
7,0,37
8,8,37
9,0,37
0,0,38
1,10,38
2,4,38
3,0,38
4,0,38
5,2,38
6,2,38
7,0,38
8,3,38
9,0,38
0,0,39
1,3,39
2,9,39
3,0,39
4,0,39
5,8,39
6,8,39
7,0,39
8,4,39
9,0,39
0,0,40
1,10,40
2,10,40
3,0,40
4,0,40
5,9,40
6,2,40
7,0,40
8,6,40
9,0,40
0,0,41
1,3,41
2,4,41
3,0,41
4,0,41
5,0,41
6,4,41
7,0,41
8,7,41
9,0,41
0,0,42
1,10,42
2,1,42
3,0,42
4,0,42
5,4,42
6,6,42
7,0,42
8,0,42
9,0,42
0,0,43
1,6,43
2,4,43
3,0,43
4,0,43
5,10,43
6,8,43
7,0,43
8,2,43
9,0,43
0,0,44
1,9,44
2,3,44
3,0,44
4,0,44
5,9,44
6,0,44
7,0,44
8,4,44
9,0,44
0,0,45
1,3,45
2,3,45
3,0,45
4,0,45
5,0,45
6,5,45
7,0,45
8,5,45
9,0,45
0,0,46
1,4,46
2,7,46
3,0,46
4,0,46
5,0,46
6,10,46
7,0,46
8,2,46
9,0,46
0,0,47
1,0,47
2,3,47
3,0,47
4,0,47
5,0,47
6,4,47
7,0,47
8,1,47
9,0,47
0,0,48
1,10,48
2,5,48
3,0,48
4,0,48
5,3,48
6,5,48
7,0,48
8,8,48
9,0,48
0,0,49
1,6,49
2,7,49
3,0,49
4,0,49
5,6,49
6,8,49
7,0,49
8,9,49
9,0,49
0,0,50
1,4,50
2,7,50
3,0,50
4,0,50
5,2,50
6,7,50
7,0,50
8,10,50
9,0,50
0,0,51
1,1,51
2,0,51
3,0,51
4,0,51
5,10,51
6,7,51
7,0,51
8,7,51
9,0,51
0,0,52
1,9,52
2,0,52
3,0,52
4,0,52
5,9,52
6,6,52
7,0,52
8,2,52
9,0,52
0,0,53
1,6,53
2,10,53
3,0,53
4,0,53
5,2,53
6,0,53
7,0,53
8,8,53
9,0,53
0,0,54
1,10,54
2,5,54
3,0,54
4,0,54
5,7,54
6,7,54
7,0,54
8,8,54
9,0,54
0,0,55
1,1,55
2,9,55
3,0,55
4,0,55
5,1,55
6,8,55
7,0,55
8,0,55
9,0,55
0,0,56
1,6,56
2,4,56
3,0,56
4,0,56
5,9,56
6,5,56
7,0,56
8,10,56
9,0,56
0,0,57
1,10,57
2,7,57
3,0,57
4,0,57
5,7,57
6,2,57
7,0,57
8,7,57
9,0,57
0,0,58
1,4,58
2,8,58
3,0,58
4,0,58
5,2,58
6,10,58
7,0,58
8,9,58
9,0,58
0,0,59
1,3,59
2,2,59
3,0,59
4,0,59
5,1,59
6,8,59
7,0,59
8,1,59
9,0,59
0,0,60
1,4,60
2,9,60
3,0,60
4,0,60
5,3,60
6,1,60
7,0,60
8,3,60
9,0,60
0,0,61
1,5,61
2,3,61
3,0,61
4,0,61
5,10,61
6,6,61
7,0,61
8,6,61
9,0,61
0,0,62
1,5,62
2,0,62
3,0,62
4,0,62
5,5,62
6,5,62
7,0,62
8,3,62
9,0,62
0,0,63
1,8,63
2,4,63
3,0,63
4,0,63
5,7,63
6,1,63
7,0,63
8,7,63
9,0,63
0,0,64
1,1,64
2,5,64
3,0,64
4,0,64
5,0,64
6,9,64
7,0,64
8,2,64
9,0,64
0,0,65
1,2,65
2,8,65
3,0,65
4,0,65
5,1,65
6,10,65
7,0,65
8,1,65
9,0,65
0,0,66
1,1,66
2,6,66
3,0,66
4,0,66
5,8,66
6,9,66
7,0,66
8,0,66
9,0,66
0,0,67
1,2,67
2,5,67
3,0,67
4,0,67
5,6,67
6,6,67
7,0,67
8,7,67
9,0,67
0,0,68
1,8,68
2,0,68
3,0,68
4,0,68
5,10,68
6,0,68
7,0,68
8,8,68
9,0,68
0,0,69
1,8,69
2,3,69
3,0,69
4,0,69
5,9,69
6,9,69
7,0,69
8,10,69
9,0,69
0,0,70
1,0,70
2,1,70
3,0,70
4,0,70
5,9,70
6,0,70
7,0,70
8,7,70
9,0,70
0,0,71
1,3,71
2,9,71
3,0,71
4,0,71
5,9,71
6,5,71
7,0,71
8,8,71
9,0,71
0,0,72
1,4,72
2,10,72
3,0,72
4,0,72
5,5,72
6,6,72
7,0,72
8,10,72
9,0,72
0,0,73
1,4,73
2,6,73
3,0,73
4,0,73
5,3,73
6,5,73
7,0,73
8,1,73
9,0,73
0,0,74
1,8,74
2,8,74
3,0,74
4,0,74
5,0,74
6,5,74
7,0,74
8,7,74
9,0,74
0,0,75
1,4,75
2,6,75
3,0,75
4,0,75
5,10,75
6,0,75
7,0,75
8,5,75
9,0,75
0,0,76
1,3,76
2,1,76
3,0,76
4,0,76
5,0,76
6,5,76
7,0,76
8,1,76
9,0,76
0,0,77
1,6,77
2,6,77
3,0,77
4,0,77
5,5,77
6,10,77
7,0,77
8,2,77
9,0,77
0,0,78
1,0,78
2,9,78
3,0,78
4,0,78
5,8,78
6,9,78
7,0,78
8,9,78
9,0,78
0,0,79
1,0,79
2,8,79
3,0,79
4,0,79
5,3,79
6,10,79
7,0,79
8,0,79
9,0,79
0,0,80
1,10,80
2,7,80
3,0,80
4,0,80
5,5,80
6,5,80
7,0,80
8,9,80
9,0,80
0,0,81
1,4,81
2,9,81
3,0,81
4,0,81
5,1,81
6,2,81
7,0,81
8,2,81
9,0,81
0,0,82
1,4,82
2,2,82
3,0,82
4,0,82
5,4,82
6,4,82
7,0,82
8,4,82
9,0,82
0,0,83
1,1,83
2,4,83
3,0,83
4,0,83
5,3,83
6,8,83
7,0,83
8,9,83
9,0,83
0,0,84
1,2,84
2,1,84
3,0,84
4,0,84
5,5,84
6,0,84
7,0,84
8,5,84
9,0,84
0,0,85
1,7,85
2,4,85
3,0,85
4,0,85
5,7,85
6,5,85
7,0,85
8,7,85
9,0,85
0,0,86
1,9,86
2,4,86
3,0,86
4,0,86

Game.py
from Cylinder import *
from random import random, sample

class Game:
    # Maximum random temperature to initialize the cylinders to
    TEMP_MAX = 1000

    # total: Total number of cylinders 
    # fastSize: Number of cylinders that can go in fast storage
    def __init__(self, total, fastSize):
        self.total = total
        self.fastSize = fastSize

        # Date for the current frame
        self.date = ""

        # List of cylinders
        self.cylinders = []

        # List of fast cylinders
        self.fastCylinders = []

        # Initialize Game with cylinders with random temperatures and locations
        self.__initRandomCylinders()

        # Set min and max temperatures
        self.maxTemp = 0
        self.minTemp = 0

        self.updateColors() 

        # Have the temperatures changed since the last updateColors call
        self.haveTempsChanged = False

    # Private methods for internal use

    # Updates max and min temperatures in the cylinders array
    def __updateMinMaxTemps(self):
        self.minTemp = self.maxTemp = self.cylinders[0].getTemperature()

        for cylinder in self.cylinders:
            temperature = cylinder.getTemperature()

            if temperature > self.maxTemp: 
                self.maxTemp = temperature
            elif temperature < self.minTemp:
                self.minTemp = temperature

    # Initializes the game with cylinders with ranodm temperatures and locations
    def __initRandomCylinders(self):
        # Initialize Cylinder objects
        for i in xrange(self.total):
            randTemp = random() * self.TEMP_MAX
            newCylinder = Cylinder(randTemp, Placement.SLOW)

            self.cylinders.append(newCylinder)

        # Pick random cylinders to be placed in fast storage
        moveToFast = sample(xrange(self.total), self.fastSize)

        for i in moveToFast:
            self.cylinders[i].setPlacement(Placement.FAST)
            self.fastCylinders.append(i)
            
    # Returns a number from {1,...,10} representing the color based on the temperature
    def __computeColorBucket(self, cylinder):
        # Get relative position of temperature in terms of range [0, 1]
        pos = float(cylinder.getTemperature() - self.minTemp) / (self.maxTemp - self.minTemp)

        return int(pos * 10) # Shift to range [0, 10]

    # Public Methods

    # Updates the color values of the cylinders
    def updateColors(self):
        self.__updateMinMaxTemps()

        for cylinder in self.cylinders:
            cylinder.setColorBucket(self.__computeColorBucket(cylinder))

    # Sets the date string
    def setDate(self, date):
        self.date = date

    # Gets the date string
    def getDate(self):
        return self.date

    # Swaps the placements of the cylinders in slow/fast storage
    # i.e. If the cylinder at cylinder1Index is fast and that at cylinder2Index
    # is slow, then cylinder1Index will be slow and cylinder2Index will be fast
    def swap(self, cyl1Index, cyl2Index):
        cyl1Plce = self.cylinders[cyl1Index].getPlacement()
        cyl2Plce = self.cylinders[cyl2Index].getPlacement()

        # Update fast cylinders list if 
        if cyl1Plce == Placement.SLOW and cyl2Plce == Placement.FAST:
            self.fastCylinders.remove(cyl2Index)
            self.fastCylinders.append(cyl1Index)
        elif cyl1Plce == Placement.FAST and cyl2Plce == Placement.SLOW:
            self.fastCylinders.remove(cyl1Index)
            self.fastCylinders.append(cyl2Index)

        self.cylinders[cyl1Index].setPlacement(cyl2Plce)
        self.cylinders[cyl2Index].setPlacement(cyl1Plce)

    # Sets the temperature of a cylinder at the given index
    def setTemperatureAt(self, index, temperature):
        self.cylinders[index].setTemperature(temperature)
        self.haveTempsChanged = True


    # Gets the color bucket of the cylinder at the given index
    def getColorAt(self, index):
        if self.haveTempsChanged:
            self.updateColors()
            self.haveTempsChanged = False

        return self.cylinders[index].getColorBucket()

    # Returns a list of all the cylinders
    def getCylinders(self):
        return self.cylinders;

    # Returns a list of the indices of the fast cylinders
    def getFastCylinderIndices(self):
        return self.fastCylinders

    # Returns a list of the indices of the slow cylinders
    def getSlowCylinderIndices(self):
        return [index for index in range(self.total) if index not in self.fastCylinders]

# Testing main method
if __name__ == "__main__":
    game = Game(10, 3)

    for cylinder in game.getCylinders():
        print cylinder

    print "MaxTemp: " + str(game.maxTemp)
    print "MinTemp: " + str(game.minTemp)

    print "Fast cylinders: " + str(game.getFastCylinderIndices())
    print "Slow cylinders: " + str(game.getSlowCylinderIndices())

    print "\nSWAPPING PLACEMENTS OF INDICES 0 AND 1\n"
    game.swap(0, 1)

    for cylinder in game.getCylinders():
        print str(cylinder)

    print "Fast cylinders: " + str(game.getFastCylinderIndices())
    print "Slow cylinders: " + str(game.getSlowCylinderIndices())

    print "\nSetting cylinder at index 0's temperature to 9999"
    game.setTemperatureAt(0, 9999)

    print str(game.getCylinders()[0])

    print "Getting color of cylinder at index 0"
    print game.getColorAt(0)

    print str(game.getCylinders()[0])

NaiveRL.py
import csv
from GameBoard.GameBoard import GameBoard
from GameBoard.Colors import index_to_color
import sys
import random
import pyglet
import time


variance = 1
# each element in days is a day
# each day contains [cylinder, rate]
days = list()
all = range(10)


fast = list()
slow = list()

def parse_csv(f):
    with open(f, 'rb') as csvfile:
        reader = csv.reader(csvfile, delimiter=',',)
        count = -1
        for row in reader:
            if len(row) == 3 and int(row[2]) != count:
                days.append(dict())
                days[-1][int(row[0])] = int(row[1])
                count += 1

            elif len(row) == 3:
                days[-1][int(row[0])] = int(row[1])
            else:
                continue
    print "done parsing"

def random_init(n):
    """picks n random cylinders from all"""
    temp = all[:]
    out = list()
    for _ in range(n):
        out.append(random.choice(temp))
        temp.remove(out[-1])
    return out, temp


def mutation(fast, slow):
    fast = fast[:]
    slow = slow[:]
    def swap(i, j):
        """swaps i from fast with j from slow"""
        temp = fast[i]
        fast[i] = slow[j]
        slow[j] = temp

    for _ in range(variance):
        i = random.choice(range(len(fast)))
        j = random.choice(range(len(slow)))
        swap(i, j)
    return fast, slow


def get_colors(fast, slow, day_i):
    """returns the colors list for a particular day"""
    day = days[day_i]
    out = []
    for i in fast:
        out.append(day[i])
    for i in slow:
        out.append(day[i])
    return out

def run_test(fast, slow):
    print "trial with ", fast, slow
    names = fast+slow
    colors = get_colors(fast, slow, 0)
    color_tup = map(index_to_color, colors)

    fitness_score = 0
    day = 0
    g = GameBoard(len(all), len(fast), names, color_tup)
    g.set_date("Day: {}".format(day))

    for i in range(1, 360):
        pyglet.clock.tick()

        for window in pyglet.app.windows:
            window.switch_to()
            window.dispatch_events()
            window.dispatch_event('on_draw')
            window.flip()

        for name in names:
            index = names.index(name)
            colors_day = get_colors(fast, slow, i)
            if name in fast:
                temp = (colors_day[name] - 5)
                fitness_score += temp
            else:
                continue
                #fitness_score += -1*colors[index]+5

        day += 1
        g.update_color(days[day])
        g.set_date("Day: {}".format(day))
        g.set_score(fitness_score)
        time.sleep(0.001)
    g.close()
    print "score of:", fitness_score
    return fitness_score

def perform_tests(n):
    best_score = -sys.maxint
    first, last = None, None
    pot_first, pot_last = random_init(2)

    for i in range(n):
        print "\ntest:", i+1

        score = run_test(pot_first, pot_last)
        print best_score, score
        if score > best_score:
            print "here"
            first, last = pot_first, pot_last
            best_score = score

        print first, last
        print pot_first, pot_last
        pot_first, pot_last = mutation(first, last)


        print "after"
        print first, last
        print pot_first, pot_last


    print "best score: ", best_score
    print first, last




if __name__ == "__main__":
    parse_csv('history.txt')
    print all
    perform_tests(20)


Cylinder.py
# Enumeration class for the storage location of the Cylinders
class Placement():
    FAST = 1
    SLOW = 2
    NONE = 2

# Cylinder class storing data of a cylinder; different from UICylinder handling UI representation
class Cylinder:
    def __init__(self, temperature, placement = Placement.NONE):
        self.temperature = temperature
        self.placement = placement
        self.bucket = -1

    # Gets the current temperature of this Cylinder
    def getTemperature(self):
        return self.temperature

    # Sets the current temperature of this Cylinder
    def setTemperature(self, temperature):
        self.temperature = temperature

    # Takes in number 1 to 10 representing color. 1 is strongest Blue, 10 is strongest Red
    # Called by Game class
    def setColorBucket(self, bucket):
        self.bucket = bucket

    # Returns number 1 to 10 representing color. 1 is strongest Blue, 10 is strongest Red
    # Called by UICylinder class
    def getColorBucket(self):
        return self.bucket

    # Returns the current storage location of this cylinder
    def getPlacement(self):
        return self.placement

    # Sets the storage location of this cylinder
    def setPlacement(self, placement):
        self.placement = placement

    def __str__(self):
        return "Temperature: {}, Placement {}, bucket {}".format(self.temperature, self.placement, self.bucket)

# Tester method
if __name__ == "__main__":
    cyl1 = Cylinder(30)
    print "Temperature: " + str(cyl1.getTemperature())
    print "Placement: " + str(cyl1.getPlacement())

    cyl1.setPlacement(Placement.FAST)
    cyl1.setColorBucket(9)

    print "Placement: " + str(cyl1.getPlacement())
    print "Color Bucket: " + str(cyl1.getColorBucket())


GameBoard/UI_Cylinder.py
import pyglet

class UI_Cylinder:
    def __init__(self, x, y, name, color):
        self.length=20
        self.name = name
        self.color = color
        self.vertex = pyglet.graphics.vertex_list(4, ('v2i', (x, y,
                                                      x, y+self.length,
                                                      x+self.length, y+self.length,

                                                      x+self.length, y)),
                                                     ('c3B', color*4))

        self.vertex.draw(pyglet.gl.GL_QUADS)

        self.label = pyglet.text.Label(str(self.name),
                                       font_name="Arial",
                                       x=x+self.length/2,
                                       y=y+self.length/2,
                                       anchor_x="center",
                                       anchor_y="center")
        self.label.draw()

    def get_attributes(self):
        return self.name, self.color

    def change_color(self, color):
        self.vertex.colors[:] = color*4
        self.vertex.draw(pyglet.gl.GL_QUADS)
        self.label.draw()

    def change_name(self, name):
        self.vertex.draw(pyglet.gl.GL_QUADS)
        self.label.text = str(name)
        self.label.draw()

    def change_color_name(self, color, name):
        self.vertex.colors[:] = color*4
        self.vertex.draw(pyglet.gl.GL_QUADS)
        self.label.text = str(name)
        self.label.draw()

GameBoard/GameBoard.py
import pyglet
from UI_Cylinder import UI_Cylinder
from Colors import *
import time
import random


class GameBoard(pyglet.window.Window):

    def __init__(self, num, numFast, names, colors, length=600):
        super(GameBoard, self).__init__(length, length)
        self.length = length
        self.num = num
        self.numFast = numFast
        self.cylinders = list()
        self.score = 0
        self.score_label = None
        self.date_label = None
        self.date = None
        self.colors = colors
        self.names = names


    def on_draw(self):
        top_color = GREY + WHITE + WHITE + GREY
        pyglet.graphics.draw(4, pyglet.gl.GL_QUADS, ('v2i', (0, self.length*2/3,
                                                             0, self.length,
                                                             self.length, self.length,
                                                             self.length, self.length*2/3)),
                                                    ('c3B', top_color))

        bottom_color = BLACK + DARK_GREY + DARK_GREY + BLACK
        pyglet.graphics.draw(4, pyglet.gl.GL_QUADS, ('v2i', (0, 0,
                                                             0, self.length*2/3,
                                                             self.length, self.length*2/3,
                                                             self.length, 0)),
                                                    ('c3b', bottom_color))

        pyglet.text.Label("Fast", font_name="Arial", font_size=16, x=10, y=self.length*2/3+5).draw()
        pyglet.text.Label("Slow", font_name="Arial", font_size=16, x=10, y=self.length*2/3-25).draw()

        self._init_rectangles()
        self.update_score()
        self.update_date()

    def update_date(self):
        if not self.date:
            return
        date = self.date
        pyglet.graphics.draw(4,
                             pyglet.gl.GL_QUADS,
                             ('v2i', (0, self.length - 22,
                                      0, self.length,
                                      self.length, self.length,
                                      self.length, self.length - 22)),
                             ('c3B', WHITE * 4))
        self.score_label.draw()
        self.date_label = pyglet.text.Label(date,
                                          font_name="Arial",
                                          font_size=16,
                                          x=self.length,
                                          y=self.length-20,
                                          anchor_x='right')
        self.date_label.set_style('color', (0, 0, 0, 255))
        self.date_label.draw()

    def _init_rectangles(self):
        new_cylinders = list()

        max_per_row = 1
        while max_per_row*30 < self.length-200:
            max_per_row += 1

        row_width = 30 * max_per_row
        center = self.length/2

        fast_rows = (self.numFast / max_per_row)
        if self.numFast % max_per_row:
            fast_rows += 1

        slow_rows = (self.num - self.numFast) / max_per_row
        if (self.num - self.numFast) % max_per_row:
            slow_rows += 1

        #print "fast, slow rows", fast_rows, slow_rows

        top = self.length - 22
        middle = self.length * 2 /3
        bottom = 0

        #print "top, middle, top", top, middle, bottom

        top_diff = (top-middle) / fast_rows
        bottom_diff = (middle-bottom) / slow_rows

        #print "top_diff, bottom_diff", top_diff, bottom_diff

        if top_diff < 30 or bottom_diff < 30:
            print "error in creating board\nboard is too small"

        for i in range(self.num):
            if i < self.numFast:
                curr_row = i / max_per_row + 1

                x = center - row_width/2 + 30*(i%max_per_row)
                y = middle + (top_diff * curr_row) / 2 - 15
                temp = UI_Cylinder(x, y, self.names[i], self.colors[self.names[i]])
                new_cylinders.append(temp)
            else:

                k = i - self.numFast
                curr_row = k / max_per_row + 1
                x = center -row_width/2 + 30*(k%max_per_row)
                y = bottom + (bottom_diff * curr_row) /2 -15

                temp = UI_Cylinder(x, y, self.names[i], self.colors[self.names[i]])
                new_cylinders.append(temp)


        self.cylinders = new_cylinders

    def update_score(self):

        pyglet.graphics.draw(4,
                             pyglet.gl.GL_QUADS,
                             ('v2i', (0, self.length -22,
                                      0, self.length,
                                      self.length, self.length,
                                      self.length, self.length-22)),
                             ('c3B', WHITE*4))

        self.score_label = pyglet.text.Label("Score: {}".format(self.score),
                                             font_size=16,
                                             x=10,
                                             y=self.length-20)
        self.score_label.set_style('color', (0, 0, 0, 255))
        self.score_label.draw()
        if self.date_label:
            self.date_label.draw()

    def swap(self, i, j):
        print "updating", i, j

        if i == j:
            pass

        i_index = self.names.index(i)
        j_index = self.names.index(j)
        i_color = self.colors[i_index]

        #self.colors[i_index] = self.colors[j_index]
        #self.colors[j_index] = i_color

        self.names[i_index] = j
        self.names[j_index] = i

        return

    def update_color(self, d):
        # d maps name to color(1-10)

        for name, color_index in d.items():
            k = self.names.index(name)
            self.colors[k] = index_to_color(color_index)

    def set_score(self, score):
        self.score = score

    def set_date(self, date):
        self.date = date

if __name__ == "__main__":
    colors = list()
    colors.append(RED)
    for _ in range(19):
        colors.append(BLUE)
    names = range(50)

    colors = [random.choice(COLOR_GRADIENT) for _ in range(50)]
    g = GameBoard(50, 20, names, colors)
    count = 0

    while True:
        pyglet.clock.tick()

        for window in pyglet.app.windows:
            window.switch_to()
            window.dispatch_events()
            window.dispatch_event('on_draw')
            window.flip()

        x = random.choice(range(50))
        y = random.choice(range(50))
        g.swap(x, y)

        count += 1
        g.set_score(count)
        time.sleep(0.4)

